src/handle_action.cpp

```diff
@@ -2380,7 +2380,8 @@ bool game::do_regular_action( action_id &act, avatar &player_character,
 
             if( has_vehicle_control( player_character ) ) {
                 const optional_vpart_position vp = get_map().veh_at( player_character.pos_bub() );
-                if( vp->vehicle().is_rotorcraft() ) {
+
+                if( vp->vehicle().is_rotorcraft() || vp->vehicle().is_abstracted_aircraft() ) {
```

Comment:
You have quite blatantly searched for every instance of is_rotorcraft() and simply changed it to an OR check for is_abstracted_aircraft().

In what way are these abstracted aircraft meant to differ from rotorcraft? Why do they function exactly the same?

Reply:
Because I did a bunch of research on fictional aircraft to see how would the fly-grantiing vehicle part I want XE's inventors to be able to craft work in flight, and every single example that wasn't based on real-world aircraft proved to be incredibly similar in how they behaved despite their wildly varying designs.

Said behavior is exactly as how a hypothetical rotorless helicopter would work, so the two classes function in similar ways:
- Moving vertically can be done without horizontal movement.
- No need to move to stay in the air.
- There's a need to turn to face the direction the driver want to go forward.
- Rotating the vehicle in flight is impacted by the vehicle's speed.
- Acceleration is quick but not instant and ground distance doesn't noticeably impact it.
- It is affected by air resistance as regular vehicles are.
- Power/fuel efficiency is dictated by the engines.
- Flight comes from one or multiple relatively small parts, placed inside or outside the vehicle.
- As long as said parts aren't compromised directly or indirectly the aircraft still flies.
- The parts can work at less than their full power.
- Destroying one of said parts causes the aircraft to crash to the ground.
- Etc.

The abstracted_aircraft class differs from the rotorcraft class as it represents the kind of flying vehicles that, despite working through literal magic or similar science, all fly the same way and can be represented by a single system so mods won't have to add a new vehicle class for every new fictional aircraft they'll decide to add. Rotorcraft is too inflexible for this purpose.

I added the multiple OR as the checks for flying vehicles require allowing a class to fly and that helicopters need their rotor checks to work properly so a new class allows to not imperatively need rotors without compromising how helicopters work. There is also a sufficient overlap in their behavior that writting the code twice would be redundant.

Reply:
> and every single example that wasn't based on real-world aircraft proved to be incredibly similar in how they behaved despite their wildly varying designs.

Then you haven't searched in the right places. Counterpoint: UFOs, flying brooms, magic magic flying boats, microteleport drives, Santa's reindeer.

 + real-life spacecraft don't need to turn to face the movement direction (you need a thruster on the opposite side of where the direction of force, but only mass concerns are preventing you from adding multiple thrusters, and direction of force is not necessarily the direction of motion). In principle, the same could be done for aircraft.  (And in reality, this is done to a mall degree with differential thrust in winged aircraft IIUC, though this isn't an exact fit.)
+ In fact, this has been done IRL with flying saucer research in the US military, with their hovering flying saucers (see: Avrocar) (I'm not sure if they actually did use such a design, but https://upload.wikimedia.org/wikipedia/commons/e/e6/Avrocar_schematic_%28high_resolution%29.jpg surely looks like they _could_ have if they added some cameras for the pilot and some extra openable/closeable holes for airflow).
+ Ground effect vehicles (like the Avrocar?) can't go high in the air, because then there is no ground effect anymore.
+ Flying brooms in HP don't work on the the same principles -- they seem to be able to orient themselves in any direction (unlike rotor aircraft, I think), have no control surfaces, do not emit hot air, in some versions they "function of Aristotelian physics instead of Newtionian physics" (whatever that's supposed to mean) (I don't know if that's canon but that doesn't matter here), IIRC they don't need energy to hover (but presumably (unconfirmed) need energy from the user to move)
+ I'm not an expert in MTG, but going by https://mtg.fandom.com/wiki/Skyship_Weatherlight it doesn't at all fly like a rotorcraft, it seems to fly somewhere in-between a sailboat (sails), fixed-wing aircraft (airfoil configuration) and zeppelin (levitation, except it requires power and not just to heat the air)
 (its "levitation field" is provided by a magic crystal taking mana as fuel, other movement is by sailing -- presumably very different energy efficiencies).
+ the Sparkle Drive in the Maretian (see chapter [Sol 6](https://www.fimfiction.net/story/396744/2/the-maretian/sol-6)): it's a setting where teleportation is possible, where shorter jumps require exponentially less power, so it frequently makes many small teleports. (It's actually a spacecraft, but an atmospheric variant could easily be imagined -- maybe magic can take care of the air that is in the way). I don't know _how_ power requirement
+ giant (flying) mecha physics (not just armored suits, _giant_ mecha). Idk _how_ they are supposed to work, but even if you ignore the internals and propulsion method, they don't seem to care about physics -- it seems to function on the power of friendship(tm), guts and looking cool instead of silly outdated notions like the laws of thermodynamics, aerodynamics and square-cube laws. Idk how to quantify such mecha, but surely they aren't like rotorcraft.
+ Santa's reindeer can fly, or move over ground, but can't hover or do vertical liftoff (in my version).

> Rotating the vehicle in flight is impacted by the vehicle's speed.

Doesn't matter (as much) for rotationally symmetric designs, just put the inside in a gyroscope so the inside can rotate independently of the outside. Mechanically it might be complicated (especially to get the air in/out of the engine), but it seem all theoretically possible, would even work _without_ magic. 

> Power/fuel efficiency is dictated by the engines.

Yes, but _how_ (quantitively) is it dictated by the engines? Aerodynamics is complicated. If you just add an engine at an arbitrary spot without regards to aerodynamics, it's probably going to be much less efficient. You need to quantify it, instead of only a qualitative comparison. Likewise for acceleration.

> Flight comes from one or multiple relatively small parts, placed inside or outside the vehicle.

In all cases I know of, flight comes from _large_ parts (in jetpack, the whole thing provides flight) (in anything resembling aircraft, the shape of the whole thing provides flight -- lifting surfaces, control surfaces, aerodynamics -- the engines theirselves sometimes appear to be an afterthought, and doesn't provide flight on its own unless you like an uncontrollable missile)), in UFOs a large part is dedicated to engines (see e.g. avrocar schematic), in flying sailboats you have large sails, sometimes the power source or engine is a big thing _around_ the craft (see e.g. elemental airships -- I dpn't know if those rings are only the power source or also the engine: https://www.pinterest.se/pin/401735229261216889/).
 
> Destroying one of said parts causes the aircraft to crash to the ground.

Santa's sleigh won't necessarily crash from taking out out a single reindeer (assuming it's pulled by multiple) -- it might need an emergency landing, but not a crash.

In the levitating sailboats, destroying a sail wouldn't stop levitation.

If it's like a fixed-wing aircraft, with some luck, even if the engine is taken out it might still function as a glider. You might even land safely with one wing missing: https://theaviationist.com/2014/09/15/f-15-lands-with-one-wing.

> No need to move to stay in the air.

Fixed-wing aircraft (I don't know any magic examples, but I imagine science fiction variants exist at least).
Santa's reindeer.
Steampunk wings.

> The parts can work at less than their full power.

But how does _efficiency_ scale, and how does power translate in _movement_.

>It is affected by air resistance as regular vehicles are.

In Star Trek, some cloaking system can go _through_ matter (is it really invisibility if you are still visible by atomic repulsion?). Note that air is a form of matter. If such a cloaking system was installed in an aircraft, it wouldn't be affected by air resistance. This creates new problems, but let's say that when stopping, the air is magically removed at cost of some power, and that it moves purely by propulsion instead of aerodynamics.

> The abstracted_aircraft class differs from the rotorcraft class as it represents the kind of flying vehicles that, despite working through literal magic or similar science, all fly the same way and can be represented by a single system so mods won't have to add a new vehicle class for every new fictional aircraft they'll decide to add. Rotorcraft is too inflexible for this purpose.

Name doesn't match purpose. If in almost all aspects, it's rotorcraft, it's not 'abstracted aircraft', because other forms of aircraft exist (e.g. CDDA has fixed-wing planes, although you can't fly them). If all you implement is rotor-like aircraft instead of more general aicraft, you shouldn't give it a lying name. E.g. call it `rotorlike_aircraft` instead of `abstracted_aircraft`, or something like that.

Also, since not all fictional aircraft behaves like rotorcraft and you only provide a method to add rotor-like aircraft, mods still need to add new vehicle classes (or add optional behaviour to old vehicle classes, or make classes modifiable through JSON code, whatever) to support their non-rotor aircraft.